tree/ntuple/v7/src/RNTupleIndex.cxx

```diff
@@ -0,0 +1,103 @@
+/// \file RNTupleIndex.cxx
+/// \ingroup NTuple ROOT7
+/// \author Florine de Geus <florine.de.geus@cern.ch>
+/// \date 2024-04-02
+/// \warning This is part of the ROOT 7 prototype! It will change without notice. It might trigger earthquakes. Feedback
+/// is welcome!
+
+/*************************************************************************
+ * Copyright (C) 1995-2024, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#include <ROOT/RNTupleIndex.hxx>
+
+ROOT::Experimental::Internal::RNTupleIndex::RNTupleIndex(std::vector<std::unique_ptr<RFieldBase>> &fields,
+                                                         RPageSource &pageSource)
+   : fFields(std::move(fields))
+{
+   std::vector<RFieldBase::RValue> fieldValues;
+   fieldValues.reserve(fields.size());
+   for (const auto &field : fFields) {
+      fieldValues.emplace_back(field->CreateValue());
+   }
+
+   std::vector<void *> ptrs;
+   ptrs.reserve(fieldValues.size());
+   for (std::uint64_t i = 0; i < pageSource.GetNEntries(); ++i) {
+      ptrs.clear();
+      for (auto &fieldValue : fieldValues) {
+         fieldValue.Read(i);
+         ptrs.push_back(fieldValue.GetPtr<void>().get());
+      }
+      Add(ptrs, i);
+   }
+}
+
+void ROOT::Experimental::Internal::RNTupleIndex::Add(const std::vector<void *> &valuePtrs, NTupleSize_t entry)
+{
+   RIndexValue indexValue;
+   for (unsigned i = 0; i < fFields.size(); ++i) {
+      indexValue += fFields[i]->GetHash(valuePtrs[i]);
+   }
+   fIndex[indexValue].push_back(entry);
+}
+
+ROOT::Experimental::NTupleSize_t
+ROOT::Experimental::Internal::RNTupleIndex::GetEntryIndex(const std::vector<void *> &valuePtrs) const
+{
+   auto entryIndices = GetEntryIndices(valuePtrs);
+   if (entryIndices.empty())
+      return kInvalidNTupleIndex;
+   return entryIndices.front();
+}
+
+std::vector<ROOT::Experimental::NTupleSize_t>
+ROOT::Experimental::Internal::RNTupleIndex::GetEntryIndices(const std::vector<void *> &valuePtrs) const
+{
+   RIndexValue indexValue;
+   for (unsigned i = 0; i < fFields.size(); ++i) {
+      indexValue += fFields[i]->GetHash(valuePtrs[i]);
+   }
+
+   if (!fIndex.count(indexValue))
+      return {};
+
+   return fIndex.at(indexValue);
```

Comment:
Given that here we're not constructing the returned vector, but rather copying an existing one, we might want to return a `const vector<> *` instead.
But this depends on 3 factors which I'm not aware of, so I'll let you decide what's best:
1. how big can this vector get? If it's just a few values, the copy is no problem at all.
2. how hot will this method be?
3. if later we will want to change the implementation so that we actually need to construct a vector rather than just referencing it, how big of a problem would that be?

Reply:
What would be the reason to return a const pointer rather than a const ref in this case?

> how big can this vector get? If it's just a few values, the copy is no problem at all.

Very likely it will only contain one element. Maybe in some exotic cases a couple of them (depending on pileup), but that will definitely not be a common case.

> how hot will this method be?
Since it's called by `GetFirstEntryNumber` it will be quite hot (i.e. once every event in the event loop). Possibly `GetFirstEntryNumber` can be refactored to not use this one.

> if later we will want to change the implementation so that we actually need to construct a vector rather than just referencing it, how big of a problem would that be?
I think not a problem at all, based the points you raised I think it makes sense to change it to a reference.